### ISSUE TITLE
BUG: new reference is not decreased due to the `return NULL` statement in macro `import_array` (static analyzer reports)

### ISSUE THREAD

**Issue by Snape3058** [scipy/scipy#16231]
Due to the `return NULL` statement in macro `import_array` provided by NumPy, the following new references are leaked.

https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/integrate/_odepackmodule.c#L830
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/integrate/_quadpackmodule.c#L34
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/interpolate/src/_fitpackmodule.c#L1555
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/ndimage/src/nd_image.c#L1184
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/optimize/_lsap_module.c#L121
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/optimize/_minpackmodule.c#L34
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/signal/_sigtoolsmodule.c#L1418
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/signal/_splinemodule.c#L559
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/sparse/sparsetools/sparsetools.cxx#L570
https://github.com/scipy/scipy/blob/d5e46add54425278eb7022e4d49f92e02b96ce8c/scipy/spatial/src/distance_wrap.c#L855

Internal Report IDs: 0bfe48 77bbba a43576 3eb20e e2207c be4614 24e435 076730 da7180 bb9ac7

**Comment by rgommers** [comment]
Thanks for all the reports @Snape3058. What static analyzer are you running? And is this a regression, or is it the first time you ran your analysis?

**Comment by rgommers** [comment]
Regarding this particular report, I'm not sure that is easy to change - this is idiomatic use, as shown in for example https://numpy.org/devdocs/user/c-info.ufunc-tutorial.html#example-numpy-ufunc-for-one-dtype. This is during extension module initialization, so if that fails there are larger problems than a leaking reference.

**Comment by mckib2** [comment]
Is it possible to flip the order?  Quickly googling around I found [this guidance](https://pythonextensionpatterns.readthedocs.io/en/latest/cpp_and_numpy.html#initialising-numpy-in-a-cpython-module) proposing initializing numpy first and then only after that success is asserted creating the module:
```cpp
    assert(! PyErr_Occurred());
    import_numpy(); // Initialise Numpy
    if (PyErr_Occurred()) {
        return NULL;
    }
    ...
    return PyModule_Create(&spammodule);
```

Granted that is for `import_numpy()`, but I assume it should work for `import_array()` as well (the error checking appears to be the same for both)

**Comment by Snape3058** [comment]
@rgommers 

> What static analyzer are you running?

It is a new experimental checker for Clang Static Analyzer. And it is a part of my research that started in 2019 and has not been published yet.

> And is this a regression, or is it the first time you ran your analysis?

The reports are from the first time I ran the analyzer on this project. Although these reports were legacy reports a couple of months ago, I have confirmed that such problems still exist in the latest version.

> I'm not sure that is easy to change

Since I also found similar bugs in NumPy [(issue 21532)](https://github.com/numpy/numpy/issues/21532), I definitely know it is difficult and of low priority to fix such bugs. However, as an analyzer vendor and to satisfy my paper reviewers, I have to submit them to prove the effectiveness of my research.

I think the suggestion from @mckib2 seems to be a good idea and worth trying, as there are no data dependencies between `import_array` and the module you are initializing. Maybe we can also use this solution to fix the bugs in NumPy and its programming guidelines if it works for this project.

> This is during extension module initialization, so if that fails there are larger problems than a leaking reference.

The developers of NumPy also said this. However, module importing can be wrapped with a try block, and a failed importing does not always means program termination (e.g. try an alternative module). Therefore, I think although it is of low priority, it is still worth fixing.

**Comment by rgommers** [comment]
Thanks for the context @Snape3058. Good luck with publishing and upstreaming into Clang.

> I think the suggestion from @mckib2 seems to be a good idea and worth trying, as there are no data dependencies between `import_array` and the module you are initializing. Maybe we can also use this solution to fix the bugs in NumPy and its programming guidelines if it works for this project.

Yes I agree, that should work.

**Comment by rgommers** [comment]
> I definitely know it is difficult and of low priority to fix such bugs. However, as an analyzer vendor and to satisfy my paper reviewers, I have to submit them to prove the effectiveness of my research.

In case it helps with your reviewers, let me just say that they weren't all that difficult to fix, and they were all valid bugs that are worth fixing.

**Comment by Snape3058** [comment]
@rgommers Thanks a lot. I will try my best to edit my paper and make it get published asap.

**Comment by wzrycj** [comment]
@Snape3058 Hi. Could you kindly share a link or the title of ur paper if it has been published? I am very interested in ur work. Your research sounds fascinating!

**Comment by Snape3058** [comment]
@wzrycj I am very glad that you like my work.
Paper pre-print: http://lcs.ios.ac.cn/~maxt/PyRefcon/ASE-2023.pdf (accepted at ASE-2023, not published yet)
Experimental data: http://lcs.ios.ac.cn/~maxt/PyRefcon
